docs/3.1/docs/tsa-arch/sdlc.rst

```diff
@@ -0,0 +1,174 @@
+     * Minimize public workflows and store internal workflows in separate DAR files to allow more flexibility in adapting to evolving business needs.
+     * Separate interface definitions in their own package for better workflow management.
+  3. Test Code Separation
+     Unit tests for Daml smart contract workflows are written using Daml Script, which is compiled into DAR files. These DAR files are for testing purposes only and should not be deployed to participant nodes. Ensure test code is not mixed with production code by using separate DAR files for Daml Script testing purposes.
+
+3. Test
+=======
+
+3.1 Testing Pyramid
+-------------------
+Testing Daml apps is similar to testing other systems: prioritize automation and test at the lowest level for speed and efficiency. However, building robust Daml apps comes with specific recommendations.
+
+.. image:: images/testing-pyramid.png
+   :alt: Testing pyramid as described below.
+   :align: center
+
+* Unit Tests
+  * Use Daml Script for white-box unit tests.
+  * Mock backends and ledgers for frontend testing.
+* Integration Tests
+  * Backend: Use white-box integration tests for internal APIs that are only used by clients under the app provider’s control.
+  * Public APIs: Use black-box behavioral tests interacting at system boundaries.
+  * Test isolation: Use long-running Canton instances to avoid repeatedly paying Canton’s startup cost, and isolate tests using unique participant users and parties for each test run. One approach is appending a test run ID as a suffix to party and user names in your test harness.
+* End-to-End Tests
+  * Test workflows between end-users and systems across multiple participant nodes, backends, and frontends.
+  * Use tools like Selenium or Playwright for browser session orchestration.
+  * Test isolation: Either bootstrap the entire system for each test run or use a long-running system instance to specific tests. The latter approach supports faster test execution and quicker iterations.
+
+3.2 Flaky Tests and Time Dependencies
+-------------------------------------
+* Eliminate flaky tests to ensure developer productivity. Writing robust tests for distributed systems is challenging due to data propagation delays and concurrent execution. It is essential to identify and address flaky tests–those that fail inconsistently and incorrectly. 
+* For time-sensitive workflows, use the “passTime” function in Daml and configure reduced wait-times for faster CI execution. Workflows that incorporate calendar or time functions in their logic, such as bond lifecycling with coupon payments, can be tested by advancing time with “passTime.” For end-to-end tests, configure workflows to advance in milliseconds to reduce CI execution time. Pause and resume automation from the test harness to prevent race conditions.
```

Comment:
```suggestion
* For time-sensitive workflows, use the “passTime” function in Daml and configure reduced wait times for faster CI execution. Workflows with calendar or time functions in their logic, such as bond lifecycling with coupon payments, can be tested by advancing time with “passTime.” For end-to-end tests, configure workflows to advance in milliseconds to reduce CI execution time. Pause and resume automation from the test harness to prevent race conditions.
```